### PR TITLE
Generate All Transaction Types and TransactionReceipt Types in BlockDataGenerator

### LINKED ISSUE #1667: Fix Brittle Transaction Receipt Deserialization Logic
https://github.com/hyperledger/besu/pull/1645#discussion_r535966483

### PR DESCRIPTION
**Pull request by RatanRSur**
### PR description
Since we're going to have new transaction types, we should use all the types in our `BlockDataGenerator` by default to make sure we're getting coverage.

### Fixed Issue(s)
fixes #1667

### Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: besu/src/test/java/org/hyperledger/besu/PrivacyReorgTest.java [hunk 1]
```diff
@@ -92,9 +93,9 @@
       Bytes.fromBase64String("A1aVtMxLCUHmBVHXoZzzBgPbW/wj5axDpW9X8l91SGo=");
 
   private static final String FIRST_BLOCK_WITH_NO_TRANSACTIONS_STATE_ROOT =
-      "0x1bdf13f6d14c7322d6e695498aab258949e55574bef7eac366eb777f43d7dd2b";
+      "0xc9dcaffbebc7edc20839c80e706430a8fa885e8f703bb89f6e95633cc2b05d4d";
   private static final String FIRST_BLOCK_WITH_SINGLE_TRANSACTION_STATE_ROOT =
-      "0x16979b290f429e06d86a43584c7d8689d4292ade9a602e5c78e2867c6ebd904e";
+      "0x66fbc6ad12ef1da78740093ea2b3362e773e510e27d8f88b68c27bfc1f4d58c8";
```
Comment (RatanRSur) [inline besu/src/test/java/org/hyperledger/besu/PrivacyReorgTest.java:98]: @lucassaldanha this test broke because I changed the BlockDataGenerator. I think it's safe to change the state roots here but I don't think the test should rely on hardcoded state roots anyways. Is there a way to determine at test time what the correct state roots should be?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/TransactionReceipt.java [hunk 2]
```diff
@@ -171,26 +173,37 @@ public void writeToWithRevertReason(final RLPOutput out) {
     writeTo(out, true);
   }
```
Comment (RatanRSur) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/TransactionReceipt.java:175]: We previously tested only `FRONTIER` receipts so roundtripping issues that were exposed are fixed here.

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/core/encoding/TransactionRLPEncoderTest.java [hunk 3]
```diff
@@ -66,17 +65,4 @@ public void encodeEIP1559TxNominalCase() {
...
-    ExperimentalEIPs.eip1559Enabled = false;
-    assertThatThrownBy(
-            () ->
-                TransactionRLPEncoder.encode(
-                    TransactionRLPDecoder.decode(RLP.input(Bytes.fromHexString(EIP1559_TX_RLP))),
-                    new BytesValueRLPOutput()))
-        .isInstanceOf(RuntimeException.class)
-        .hasMessageContaining("EIP-1559 feature flag");
-    ExperimentalEIPs.eip1559Enabled = ExperimentalEIPs.EIP1559_ENABLED_DEFAULT_VALUE;
-  }
```
Comment (RatanRSur) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/core/encoding/TransactionRLPEncoderTest.java:81]: We don't fail at deseriailization anymore. We always deserialize and reject in transaction validation based on what fork we're on

File: besu/src/test/java/org/hyperledger/besu/PrivacyReorgTest.java (hunk 1 above)
Comment (abdelhamidbakhta) [reply besu/src/test/java/org/hyperledger/besu/PrivacyReorgTest.java:98]: I think actually this is not a bad thing to rely on hardcoded state roots on tests. Because when you change something it breaks, and this is good. If you use your code to generate the assertion there is a risk of working only in close loop.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/Wei.java [hunk 4]
```diff
@@ -18,7 +18,7 @@
 
 import java.math.BigInteger;
 
-import org.apache.tuweni.bytes.Bytes32;
+import org.apache.tuweni.bytes.Bytes;
```
Comment (abdelhamidbakhta) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/Wei.java:21]: Why this change is necessary ?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/encoding/TransactionRLPEncoder.java [hunk 5]
```diff
@@ -77,8 +76,6 @@ static void encodeFrontier(final Transaction transaction, final RLPOutput out) {
   }
 
   static void encodeEIP1559(final Transaction transaction, final RLPOutput out) {
-    ExperimentalEIPs.eip1559MustBeEnabled();
```
Comment (abdelhamidbakhta) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/encoding/TransactionRLPEncoder.java:80]: Why do you remove that ?

**Review by abdelhamidbakhta — APPROVED** [review]
Left some minor comments.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/Wei.java (hunk 4 above)
Comment (RatanRSur) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/Wei.java:21]: In `TransactionRLPDecoder`, we previously had `Wei.of` of a `BigInteger`, sometimes that `BigInteger` was negative so the `Wei` was negative. When I changed it to `Wei.wrap` I needed to allow `Bytes` instead of just `Bytes32` since it could be shorter.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/encoding/TransactionRLPEncoder.java (hunk 5 above)
Comment (RatanRSur) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/encoding/TransactionRLPEncoder.java:80]: In general we're moving the transactions to not be validated at serialization time. All the checks to make sure we're not creating/deserializing transactions for a different fork are done elsewhere.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/Wei.java (hunk 4 above)
Comment (abdelhamidbakhta) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/Wei.java:21]: That makes sense
